### ISSUE TITLE
Phase 2: System Design

### ISSUE THREAD

**Issue by FairyEvilMoon** [FairyEvilMoon/-Smart-Healthcare-Management-System-KU#2]
Design complex UML Diagrams:
- Use Case Diagrams for interactions between doctors, patients, and admins.
- Class Diagrams for system structure, focusing on entities like Patient, Doctor, Appointment,
- HealthMetrics, and Prescription.
- Sequence Diagrams for critical processes like real-time health monitoring and appointment
- scheduling.
- Component Diagrams to model the subsystems (patient portal, wearable integration,
- database, machine learning module).
- Design a Predictive Model for health risk analysis.
- Define the Database Schema for patient records, appointments, and health data.

Deliverables:

- [ ] UML Diagrams.
- [ ] Predictive Model Design.
- [ ] Database Schema

**Comment by FairyEvilMoon** [comment]
![shms-use-case-diagrams](https://github.com/user-attachments/assets/6ca39de2-5dbf-4104-9766-7ead10f78da0)

![shms-class-diagram](https://github.com/user-attachments/assets/5a40ca42-c74c-4020-b154-178a73d380c0)

![shms-appointment-booking-sequence](https://github.com/user-attachments/assets/d00c7d18-ccd1-4e14-970b-39e9afbd7b48)

use diagram, class diagram, sequence diagram.
if anyone is making a better one do post

**Comment by FairyEvilMoon** [comment]
Component Diagram

![shms-component-diagram](https://github.com/user-attachments/assets/4570dc03-e717-4356-a9a2-ce8ff32a8fc2)

**Comment by FairyEvilMoon** [comment]
Health Risk Predictive Model Design
1. Objective
Design a predictive model to analyze patient health data and predict potential health risks.
2. Input Features

Age
Gender
Body Mass Index (BMI)
Blood Pressure (Systolic and Diastolic)
Heart Rate
Cholesterol Levels (Total, HDL, LDL)
Blood Glucose Level
Smoking Status (Boolean)
Alcohol Consumption (units per week)
Physical Activity Level (hours per week)
Family History of:

Heart Disease
Diabetes
Cancer


Previous Diagnoses

3. Output
The model will predict the probability of the patient developing the following conditions within the next 5 years:

Cardiovascular Disease
Type 2 Diabetes
Hypertension

4. Model Architecture
We'll use an ensemble method combining multiple machine learning algorithms:

Random Forest Classifier

Good for handling a mix of categorical and numerical data
Can capture non-linear relationships
Provides feature importance


Gradient Boosting Classifier (e.g., XGBoost)

High performance and accuracy
Can handle imbalanced datasets well


Logistic Regression

Provides easily interpretable results
Good for understanding the impact of individual features



5. Training Process

Data Preprocessing:

Handle missing values
Encode categorical variables
Normalize numerical features


Feature Selection:

Use correlation analysis and feature importance from Random Forest to select the most relevant features


Model Training:

Split data into training (70%), validation (15%), and test (15%) sets
Train each model (Random Forest, XGBoost, Logistic Regression) on the training data
Use cross-validation to tune hyperparameters


Ensemble Method:

Use soft voting to combine predictions from all three models
Weight the predictions based on each model's performance on the validation set


Model Evaluation:

Use metrics such as AUC-ROC, precision, recall, and F1-score
Evaluate on the test set to assess generalization



6. Deployment and Monitoring

Deploy the model as a microservice that can be called via API
Implement logging to track model inputs and outputs
Set up monitoring for model performance and data drift
Plan for regular retraining (e.g., quarterly) to keep the model up-to-date

7. Interpretability

Use SHAP (SHapley Additive exPlanations) values to explain individual predictions
Provide feature importance plots to show which factors most influence the predictions

8. Ethical Considerations

Ensure the training data is diverse and representative
Regularly audit the model for bias, especially regarding protected characteristics like age, gender, and race
Provide clear explanations of the model's limitations to healthcare providers

**Comment by FairyEvilMoon** [comment]
# SHMS Database Schema
## 1. Users Table
sqlCopyCREATE TABLE Users (
    UserID INT PRIMARY KEY,
    Email VARCHAR(255) UNIQUE NOT NULL,
    PasswordHash VARCHAR(255) NOT NULL,
    FirstName VARCHAR(50) NOT NULL,
    LastName VARCHAR(50) NOT NULL,
    DateOfBirth DATE NOT NULL,
    Gender ENUM('Male', 'Female', 'Other') NOT NULL,
    PhoneNumber VARCHAR(20),
    UserType ENUM('Patient', 'Doctor', 'Admin') NOT NULL,
    CreatedAt TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
    UpdatedAt TIMESTAMP DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP
);
## 2. Patients Table
sqlCopyCREATE TABLE Patients (
    PatientID INT PRIMARY KEY,
    UserID INT UNIQUE,
    EmergencyContact VARCHAR(255),
    InsuranceInfo TEXT,
    FOREIGN KEY (UserID) REFERENCES Users(UserID)
);
## 3. Doctors Table
sqlCopyCREATE TABLE Doctors (
    DoctorID INT PRIMARY KEY,
    UserID INT UNIQUE,
    Specialization VARCHAR(100),
    LicenseNumber VARCHAR(50) UNIQUE,
    FOREIGN KEY (UserID) REFERENCES Users(UserID)
);
## 4. Appointments Table
sqlCopyCREATE TABLE Appointments (
    AppointmentID INT PRIMARY KEY,
    PatientID INT,
    DoctorID INT,
    AppointmentDateTime DATETIME,
    Status ENUM('Scheduled', 'Completed', 'Cancelled') DEFAULT 'Scheduled',
    Notes TEXT,
    CreatedAt TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
    UpdatedAt TIMESTAMP DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
    FOREIGN KEY (PatientID) REFERENCES Patients(PatientID),
    FOREIGN KEY (DoctorID) REFERENCES Doctors(DoctorID)
);
## 5. MedicalRecords Table
sqlCopyCREATE TABLE MedicalRecords (
    RecordID INT PRIMARY KEY,
    PatientID INT,
    DoctorID INT,
    Date DATE,
    Diagnosis TEXT,
    Treatment TEXT,
    Medications TEXT,
    Notes TEXT,
    FOREIGN KEY (PatientID) REFERENCES Patients(PatientID),
    FOREIGN KEY (DoctorID) REFERENCES Doctors(DoctorID)
);
## 6. HealthMetrics Table
sqlCopyCREATE TABLE HealthMetrics (
    MetricID INT PRIMARY KEY,
    PatientID INT,
    RecordedAt TIMESTAMP,
    HeartRate INT,
    BloodPressureSystolic INT,
    BloodPressureDiastolic INT,
    Temperature DECIMAL(3,1),
    OxygenSaturation INT,
    BloodGlucose INT,
    Weight DECIMAL(5,2),
    Steps INT,
    SleepHours DECIMAL(3,1),
    FOREIGN KEY (PatientID) REFERENCES Patients(PatientID)
);
## 7. Prescriptions Table
sqlCopyCREATE TABLE Prescriptions (
    PrescriptionID INT PRIMARY KEY,
    PatientID INT,
    DoctorID INT,
    MedicationName VARCHAR(100),
    Dosage VARCHAR(50),
    Frequency VARCHAR(50),
    StartDate DATE,
    EndDate DATE,
    Notes TEXT,
    FOREIGN KEY (PatientID) REFERENCES Patients(PatientID),
    FOREIGN KEY (DoctorID) REFERENCES Doctors(DoctorID)
);
## 8. HealthRiskPredictions Table
sqlCopyCREATE TABLE HealthRiskPredictions (
    PredictionID INT PRIMARY KEY,
    PatientID INT,
    PredictionDate DATE,
    CardiovascularDiseaseRisk DECIMAL(5,2),
    DiabetesRisk DECIMAL(5,2),
    HypertensionRisk DECIMAL(5,2),
    ModelVersion VARCHAR(20),
    FOREIGN KEY (PatientID) REFERENCES Patients(PatientID)
);

these data schemes might be modified in the future

**Comment by FairyEvilMoon** [comment]
THE UML diagrams need to be looked at and reformated if needed

**Comment by FairyEvilMoon** [comment]
need to update and make new uml diagrams
and new datascheme
